### PR TITLE
CDAP-15524. HTTP Batch Source. Pagination. Http client implementation

### PR DESCRIPTION
**Pull request by aonischuk**
1. Implemented all pagination types. Including custom pagination (via python code)
2. Implemented http client wrapper. Which includes proxy, basic auth, oauth2 request, TSL/SSL functionality, timeout handling.

### CODE REVIEW

**Comment by albertshau** [comment]
should use mocks for the unit tests, otherwise they can break due to external changes. See https://github.com/cdapio/hydrator-plugins/blob/develop/spark-plugins/src/test/java/io/cdap/plugin/spark/test/SparkPluginTest.java#L119 for an example.

File: pom.xml [hunk 1]
```diff
@@ -75,17 +75,22 @@
...
-    <app.parents>
-      system:cdap-data-pipeline[6.0.0-SNAPSHOT,7.0.0-SNAPSHOT),system:cdap-data-streams[6.0.0-SNAPSHOT,7.0.0-SNAPSHOT)
-    </app.parents>
+    <jackson.version>2.9.9</jackson.version>
+    <junit.version>4.11</junit.version>
+    <jython.version>2.7.1</jython.version>
     <!-- this is here because project.basedir evaluates to null in the script build step -->
     <main.basedir>${project.basedir}</main.basedir>
+    <unxml.version>0.9</unxml.version>
+    <widgets.dir>widgets</widgets.dir>
```
Comment (albertshau) [inline pom.xml]: this can be removed. It's a relic from when the build was using javascript instead of the cdap maven plugin

File: pom.xml [hunk 2]
```diff
@@ -103,31 +108,13 @@
...
-    </dependency>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>13.0.1</version>
-    </dependency>
-    <dependency>
-      <groupId>io.cdap.cdap</groupId>
-      <artifactId>cdap-data-pipeline</artifactId>
+      <artifactId>cdap-formats</artifactId>
```
Comment (albertshau) [inline pom.xml:111]: I think this pulls in stuff like avro and grok. Since we're not using those, let's add exclusions for them, which will make the jar a little bit smaller.

File: pom.xml [hunk 3]
```diff
@@ -241,16 +228,100 @@
...
+      <groupId>org.python</groupId>
+      <artifactId>jython-standalone</artifactId>
+      <version>${jython.version}</version>
+    </dependency>
+
+
+    <!-- tests -->
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-metadata-spi</artifactId>
```
Comment (albertshau) [inline pom.xml]: why is this needed?

File: pom.xml [hunk 4]
```diff
@@ -241,16 +228,100 @@
...
+      <version>${cdap.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-common</artifactId>
```
Comment (albertshau) [inline pom.xml]: why is this needed?

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+  private CloseableHttpClient createHttpClient(BaseHttpSourceConfig config) throws IOException, URISyntaxException {
+    HttpClientBuilder httpClientBuilder = HttpClientBuilder.create();
+    httpClientBuilder.setSSLSocketFactory(new SSLConnectionSocketFactoryCreator(config).create());
+
+    // set timeouts
+    Long connectTimeoutMillis = TimeUnit.SECONDS.toMillis(config.getConnectTimeout());
+    Long readTimeoutMillis = TimeUnit.SECONDS.toMillis(config.getReadTimeout());
+    RequestConfig.Builder requestBuilder = RequestConfig.custom();
+    requestBuilder.setSocketTimeout(readTimeoutMillis.intValue());
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:102]: if these should be ints, config.getConnectTimeout() should also return an int. Otherwise there can be weird behavior due to overflow.

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    String requestBodyString = config.getRequestBody();
+    if (requestBodyString != null) {
+      requestBody = new StringEntity(requestBodyString, Charsets.UTF_8.toString());
+    } else {
+      requestBody = null;
+    }
+  }
+
+  public CloseableHttpResponse executeHTTP(String uri) throws IOException {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:72]: is this meant to be called just once or can it be called more than once? Please add a javadoc to describe expected usage.

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  public void close() throws IOException {
+    if (httpClient != null) {
+      httpClient.close();
+    }
+  }
+
+  private void initialize() {
+    try {
+      httpClient = createHttpClient(config);
+    } catch (IOException | URISyntaxException e) {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java]: these are two pretty different types of errors, it's better not to wrap them in a runtime exception, but to let them propagate up.

I'm also confused about why createHttpClient() throws a URISyntaxException. Which method call is throwing it? Ideally this isn't an exception we would have to deal with. It would be better if this class gets passed in URI or URL objects instead of dealing with strings, with validation of those done outside this class. 

In general, validation of user input should happen as early as possible, with classes later in the logic using classes that are guaranteed to be valid.

File: src/main/java/io/cdap/plugin/http/source/common/http/SSLConnectionSocketFactoryCreator.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    KeyStoreException, IOException, UnrecoverableKeyException {
+
+    KeyStore keystore = loadKeystore(config.getKeystoreFile(), config.getKeystoreType().name(),
+                                     config.getKeystorePassword());
+
+    String keyStorePassword = config.getKeystorePassword();
+
+    // we have to manually fall back to default keystore. SSLContext won't provide such a functionality.
+    if (keystore == null) {
+      String keyStore = System.getProperty("javax.net.ssl.keyStore");
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/http/SSLConnectionSocketFactoryCreator.java:80]: How do these properties get set?

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java [hunk 9]
```diff
@@ -15,15 +15,182 @@
...
+   * @return true if page still has elements to iterate. Otherwise it will load next page.
+   * False if no more pages to load or the page loaded has no elements.
+   */
+  protected boolean ensurePageIterable() {
+    try {
+      if (page == null || !page.hasNext()) {
+        page = getNextPage();
+      }
+
+      return (page != null && page.hasNext());
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: nit: can remove the outer parentheses

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java [hunk 10]
```diff
@@ -15,15 +15,182 @@
...
+    }
+
+    try {
+      Awaitility
+        .await().with()
+        .pollInterval(pollInterval)
+        .timeout(config.getMaxRetryDuration(), TimeUnit.SECONDS)
+        .until(this::visitPageAndCheckStatusCode);
+    } catch (ConditionTimeoutException ex) {
+      // Retries failed. We don't need to do anything here. This will ba handled using httpStatusCode below.
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: typo: ba -> be

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 9 above)
Comment (albertshau) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: this will return a false negative if it's possible to have an empty Page. To protect against this, it's better to loop above until we get a page with entries.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java [hunk 11]
```diff
@@ -15,15 +15,182 @@
...
+    }
+
+    response = httpClient.executeHTTP(nextPageUrl);
+    httpStatusCode = response.getStatusLine().getStatusCode();
+    HttpErrorHandlingStrategy errorHandlingStrategy = httpErrorHandler.getErrorHandlingStrategy(httpStatusCode);
+
+    return !errorHandlingStrategy.shouldRetry();
+  }
+
+  protected BasePage getNextPage() throws IOException, InterruptedException {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: annotate as nullabe

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java [hunk 12]
```diff
@@ -15,15 +15,182 @@
...
+    }
+
+    // response being null, means it's the first page we are loading
+    if (response != null && config.getWaitTimeBetweenPages() != null) {
+      Thread.sleep(config.getWaitTimeBetweenPages());
+    }
+
+    LOG.debug("Fetching '{}'", nextPageUrl);
+
+    PollInterval pollInterval;
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: it looks like this could be a private final that is set once in the constructor.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java [hunk 13]
```diff
@@ -15,15 +15,182 @@
...
+        .await().with()
+        .pollInterval(pollInterval)
+        .timeout(config.getMaxRetryDuration(), TimeUnit.SECONDS)
+        .until(this::visitPageAndCheckStatusCode);
+    } catch (ConditionTimeoutException ex) {
+      // Retries failed. We don't need to do anything here. This will ba handled using httpStatusCode below.
+    }
+
+    HttpEntity reponseEntity = response.getEntity();
+    byte[] responseBytes = EntityUtils.toByteArray(reponseEntity);
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:114]: we should not be loading the entire contents into memory unless it's an error. Under normal circumstances the contents should be read as an InputStream.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java [hunk 14]
```diff
@@ -15,15 +15,182 @@
...
+    if (config.getFormat().equals(PageFormat.BLOB)) {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: Format should only apply on success. Most of the time, an error response is not formatted the same way a successful response is formatted.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/JythonPythonExecutor.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+    try {
+      interpreter.exec(script);
+    } catch (PySyntaxError e) {
+      if (Strings.isNullOrEmpty(e.getMessage())) {
+        throw new RuntimeException(
+          "Unknown syntax error occurred while interpreting python code. Please make sure syntax is correct.", e);
+      }
+      throw new RuntimeException(
+        String.format("Syntax error occurred while interpreting python code: '%s'" +
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/JythonPythonExecutor.java]: Should not surround the exception message with text, as it makes it hard to distinguish between what comes from the exception and what comes from this line. Better to just have:

Syntax error occurred while interpreting python code: '%s'

File: src/main/java/io/cdap/plugin/http/source/common/pagination/LinkInResponseBodyPagionationIterator.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Every page contains a next page url. This pagination type is only supported for JSON and XML formats.
+ * Pagination happens until no next page field is present or until page contains no elements.
+ */
+public class LinkInResponseBodyPagionationIterator extends BaseHttpPaginationIterator {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/LinkInResponseBodyPagionationIterator.java]: typo: Pagio -> Pagi

File: src/main/java/io/cdap/plugin/http/source/common/pagination/NonePagionationIterator.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+import io.cdap.plugin.http.source.common.BaseHttpSourceConfig;
+import io.cdap.plugin.http.source.common.pagination.page.BasePage;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Returns a single page of data
+ */
+public class NonePagionationIterator extends BaseHttpPaginationIterator {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/NonePagionationIterator.java]: typo: Pagio -> Pagi

**Review by albertshau — COMMENTED** [review]
Can you move the format stuff to a separate PR? 2000 lines is a lot and there aren't any unit tests yet. Pagination and format shouldn't share much code.

**Comment by albertshau** [comment]
After taking a look, you also shouldn't need to setup an actual http server for the unit tests. The pagination classes can all be tested using a mock HttpClient that returns a response that you set in the test.

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java (hunk 7 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java]: URISyntaxException is thown by OAuthUtil.getAccessTokenByRefreshToken() Which forms URI for retrieving accessToken by RefreshToken.

File: src/main/java/io/cdap/plugin/http/source/common/http/SSLConnectionSocketFactoryCreator.java (hunk 8 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/http/SSLConnectionSocketFactoryCreator.java:80]: They are unset by default. But to change default keyStore user can set them via "java -D....".

This code is there just not to break default ssl settings behavior in java when javax.net.ssl.* properties are set.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 9 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: I don't think Page instance can  be null. See PageFactory#createInstance it cannot return null. Null can only be returned from getNextPage() as a sign that there are no more pages to load.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 13 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:114]: At some point we will need to read whole byte array into memory. Since it needs to be written to structured record "body" field (field type is "bytes"). Am I missing something?

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java (hunk 6 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:72]: done

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 11 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: done

File: pom.xml (hunk 1 above)
Comment (aonischuk) [reply pom.xml]: done

File: pom.xml (hunk 2 above)
Comment (aonischuk) [reply pom.xml:111]: done

File: pom.xml (hunk 4 above)
Comment (aonischuk) [reply pom.xml]: done

File: pom.xml (hunk 3 above)
Comment (aonischuk) [reply pom.xml]: done

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java (hunk 5 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java:102]: done. However note that TimeUnit.SECONDS.toMillis still returns long. So conversion to int remains.

File: src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java (hunk 7 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java]: done

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 9 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: done (removed the outer parentheses)

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 10 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: done

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 12 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: done

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 14 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: done

File: src/main/java/io/cdap/plugin/http/source/common/pagination/JythonPythonExecutor.java (hunk 15 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/JythonPythonExecutor.java]: done

File: src/main/java/io/cdap/plugin/http/source/common/pagination/NonePagionationIterator.java (hunk 17 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/NonePagionationIterator.java]: done

File: src/main/java/io/cdap/plugin/http/source/common/pagination/LinkInResponseBodyPagionationIterator.java (hunk 16 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/LinkInResponseBodyPagionationIterator.java]: done

**Comment by aonischuk** [comment]
1. Removed the format stuff from PR.
2. Regarding UT: I have added unit tests
3. Thank you for reviewing so much code lately.

@albertshau I have addressed your comments. Can we please merge if everything looks good?

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 13 above)
Comment (albertshau) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:114]: That's only for the 'blob' format. Everything else should not. You can ignore this now for now, since format has been moved out.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 9 above)
Comment (albertshau) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: Empty page meaning one where page.hasNext() is false.

If page.hasNext() is false and getNextPage() would have returned a non-null, then this will terminate early.

**Comment by albertshau** [comment]
One remaining comment, please fix and squash, then i'll merge.

**Comment by aonischuk** [comment]
@albertshau squashed the commits. Please feel free to merge.

Ps. also added pagination unit tests.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java [hunk 18]
```diff
@@ -15,15 +15,194 @@
...
+   * @return true if page still has elements to iterate. Otherwise it will load next page.
+   * False if no more pages to load or the page loaded has no elements.
+   */
+  protected boolean ensurePageIterable() {
+    try {
+      if (page == null || !page.hasNext()) {
+        page = getNextPage();
+      }
+
+      return page != null && page.hasNext();
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:165]: comment here is still not addressed.

File: src/test/java/io/cdap/plugin/http/source/common/pagination/PaginationIteratorTest.java [hunk 19]
```diff
@@ -0,0 +1,288 @@
+package io.cdap.plugin.http.source.common.pagination;
```
Comment (albertshau) [inline src/test/java/io/cdap/plugin/http/source/common/pagination/PaginationIteratorTest.java:17]: missing copyright

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 9 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java]: Empty page (with no elements) in pagination APIs usually means that you should stop paginating otherwise you'll get into endless loop.

Example API:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=80

You can go to page 100 200 500 1000 which are empty and it's endless. All with no elements.
Page 10.000:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=10000

**Comment by aonischuk** [comment]
1. Answered your remaining comment
2. Added pagination unit tests.
3. Squashed the commit.

@albertshau please feel free to merge.

**Comment by albertshau** [comment]
Did you leave an answer about my comment? Didn't see it.

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 18 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:165]: Empty page (with no elements) in pagination APIs usually means that you should stop paginating otherwise you'll get into endless loop.

Example API:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=80

You can go to page 100 200 500 1000 which are empty and it's endless. All with no elements.
Page 10.000:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=10000

Comment (aonischuk) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:165]: that's why I don't think we need any change here. Please correct me if I am wrong.

File: src/test/java/io/cdap/plugin/http/source/common/pagination/PaginationIteratorTest.java (hunk 19 above)
Comment (aonischuk) [reply src/test/java/io/cdap/plugin/http/source/common/pagination/PaginationIteratorTest.java:17]: done

**Comment by aonischuk** [comment]
1. Added copyright
2. Answered the comment regarding empty page getNext().

@albertshau please re-check/merge

File: src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java (hunk 18 above)
Comment (albertshau) [reply src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java:165]: I see, makes sense

**Comment by aonischuk** [comment]
Thanks!
